kong/db/schema/init.lua

```diff
@@ -2054,48 +2054,68 @@ end
 --- Run transformations on fields.
 -- @param input The input table.
 -- @param original_input The original input for transformation detection.
+-- @param context a string describing the CRUD context:
+-- valid values are: "insert", "update", "upsert", "select"
 -- @return the transformed entity
-function Schema:transform(input, original_input)
+function Schema:transform(input, original_input, context)
   if not self.transformations then
     return input
   end
 
   local output = input
   for _, transformation in ipairs(self.transformations) do
-    local on_write = transformation.on_write
-    local args = {}
-    local argc = 0
-    local all_set  = true
-    for _, input_field_name in ipairs(transformation.input) do
-      local value = get_field(original_input or input, input_field_name)
-      if is_nonempty(value) then
-        argc = argc + 1
-        if original_input then
-          args[argc] = get_field(input, input_field_name)
-        else
-          args[argc] = value
-        end
+    local transform
+    if context == "select" then
+      transform = transformation.on_read
 
-      else
-        all_set = false
-        break
-      end
+    else
+      transform = transformation.on_write
     end
 
-    if all_set then
-      if transformation.needs then
-        for _, need in ipairs(transformation.needs) do
+    if transform then
+      local args = {}
+      local argc = 0
+      local all_set  = true
+      for _, input_field_name in ipairs(transformation.input) do
+        local value = get_field(original_input or input, input_field_name)
+        if is_nonempty(value) then
           argc = argc + 1
-          args[argc] = get_field(input, need)
+          if original_input then
+            args[argc] = get_field(input, input_field_name)
+          else
+            args[argc] = value
+          end
+
+        else
+          all_set = false
```

Comment:
Can we avoid setting `all_set` at all? It seems we could do `return` here instead of setting `all_set` to false + break + an if a bit later on.